pkg/model/hetznermodel/servers.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hetznermodel
+
+import (
+	"strconv"
+
+	"k8s.io/kops/pkg/model"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/hetzner"
+	"k8s.io/kops/upup/pkg/fi/cloudup/hetznertasks"
+)
+
+// ServerModelBuilder configures network objects
+type ServerModelBuilder struct {
+	*HetznerModelContext
+	Lifecycle              fi.Lifecycle
+	BootstrapScriptBuilder *model.BootstrapScriptBuilder
+}
+
+var _ fi.ModelBuilder = &ServerModelBuilder{}
+
+func (b *ServerModelBuilder) Build(c *fi.ModelBuilderContext) error {
+	//sshKeyName, err := b.SSHKeyName()
+	//if err != nil {
+	//	return err
+	//}
+	//klog.Infof("SSH key name: %v", sshKeyName)
+	//for _, key := range b.SSHPublicKeys {
+	//	klog.Infof("SSH pub key: %v", string(key))
+	//}
+
+	//splitSSHKeyName := strings.Split(sshKeyName, "-")
+	//sshKeyFingerPrint := splitSSHKeyName[len(splitSSHKeyName)-1]
+
+	for _, ig := range b.InstanceGroups {
+		igSize := fi.Int32Value(ig.Spec.MinSize)
+
+		labels := make(map[string]string)
+		labels[hetzner.TagKubernetesClusterName] = b.ClusterName()
+		labels[hetzner.TagKubernetesInstanceGroup] = ig.Name
+		labels[hetzner.TagKubernetesInstanceRole] = string(ig.Spec.Role)
+
+		userData, err := b.BootstrapScriptBuilder.ResourceNodeUp(c, ig)
+		if err != nil {
+			return err
+		}
+
+		for i := 1; i <= int(igSize); i++ {
+			// hcloud-cloud-controller-manager requires hostname to be same as server name.
+			// This means server names should not contain the cluster name (which contains "." chars"
```

Comment:
You could maybe also replace the dots with dashes.  However, I like what you've done here, and it looks like we have the cluster name in the labels.

I guess one maybe-problem is if users were running two clusters in the same account, these names could collide?  You could add a random suffix if that's a concern (like k8s does for deployment pods)

Reply:
Actually was thinking of adding the nodeup config hash.
Servers names must be unique per project, so one could use the same account for multiple clusters in different projects.